File: examples/python/03-streaming-deltas.py
# 03 — Streaming deltas
#
# Subscribe to session events so each AssistantMessageData chunk prints as it
# arrives. Events fire automatically during `send_and_wait()`, so no separate
# idle-await is needed in the common case.
#
# Run: python 03-streaming-deltas.py

import asyncio
import os
import shutil
import sys

from copilot import CopilotClient, SubprocessConfig
from copilot.generated.session_events import AssistantMessageData
from copilot.session import PermissionHandler


def resolve_copilot_cli() -> str:
    """Resolve the copilot CLI: PATH first, COPILOT_CLI_PATH env as fallback."""
    path = shutil.which("copilot")
    if path:
        return path
    env_path = os.environ.get("COPILOT_CLI_PATH")
    if env_path:
        return env_path
    raise RuntimeError(
        "Copilot CLI not found. Install @github/copilot globally or set COPILOT_CLI_PATH."
    )


async def main() -> None:
    client = CopilotClient(SubprocessConfig(
        cli_path=resolve_copilot_cli(),
        cli_args=["--disable-builtin-mcps"],
    ))
    await client.start()
    try:
        session = await client.create_session(
            model="gpt-4.1",
            streaming=True,
            on_permission_request=PermissionHandler.approve_all,
        )

        def on_event(event) -> None:
            if isinstance(event.data, AssistantMessageData):
                sys.stdout.write(event.data.content or "")
                sys.stdout.flush()

        unsubscribe = session.on(on_event)
        try:
            await session.send_and_wait(
                "Write a haiku about Python's Global Interpreter Lock."
            )
        finally:
            unsubscribe()
        sys.stdout.write("\n")
    finally:
        await client.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/nodejs/simple-sdk/03-streaming-deltas.mjs
// 03 — Streaming deltas
//
// Set `streaming: true` on the session and subscribe to `assistant.message_delta`
// to print chunks as they arrive. `session.idle` fires when the response is
// complete and the session is ready for the next prompt.
//
// Run: node examples/03-streaming-deltas.mjs

import { createRequire } from "node:module";
import { CopilotClient, approveAll } from "@github/copilot-sdk";

// Resolve the @github/copilot CLI relative to this script first, fall back to
// the COPILOT_CLI_PATH env var. `--disable-builtin-mcps` keeps the example on
// the bare SDK surface; the env var COPILOT_DISABLE_MCP is not honored.
function resolveCopilotCli() {
  try {
    return createRequire(import.meta.url).resolve("@github/copilot");
  } catch {
    if (process.env.COPILOT_CLI_PATH) return process.env.COPILOT_CLI_PATH;
    throw new Error("Copilot CLI not found. Run `npm install` or set COPILOT_CLI_PATH.");
  }
}

const client = new CopilotClient({
  cliPath: resolveCopilotCli(),
  cliArgs: ["--disable-builtin-mcps"],
});
const session = await client.createSession({
  model: "gpt-4.1",
  streaming: true,
  onPermissionRequest: approveAll,
});

const offDelta = session.on("assistant.message_delta", (event) => {
  process.stdout.write(event?.data?.content ?? "");
});

const idle = new Promise((resolve) => {
  const offIdle = session.on("session.idle", () => {
    offIdle();
    resolve();
  });
});

await session.sendAndWait({
  prompt: "Write a haiku about garbage collection in JavaScript.",
});
await idle;

offDelta();
process.stdout.write("\n");

await client.stop();
process.exit(0);
